APIs for saving custom queries

pranavgore09 opened the pull request:
Allow user to save queries.
Title, SpaceID, Creator - should be unique

Create, List APIs need user to be logged in

ToDo:
 - ~~Show one query API~~
 - ~~Delete API~~
 - ~~move query verification from controller to repo~~
 - update this description with all use cases for APIs
 - ~~add the total count in meta for List API~~
 - ~~make login required for SHOW api~~

pranavgore09 commented:
[test]

pranavgore09 commented:
[test]

baijum commented:
[test]

baijum commented on controller/query.go:
```diff
@@ -0,0 +1,147 @@
+			Creator: *currentUserIdentityID,
+		}
+		// Parse fields to make sure that query is valid
+		_, _, err := search.ParseFilterString(ctx, q.Fields)
```
The `err` value only indicate whether the query is syntactically correct.  Where as the first return value is of type `criteria.Expression` and if that value is `nil`, the expression is still invalid. Should we catch that?

pranavgore09 replied:
thanks @baijum I wasn't aware of it, done in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/2c7f2d8b82dfd68ccfa159a7e6373bfe991266c4

baijum commented on controller/query.go:
```diff
@@ -0,0 +1,147 @@
+		res := &app.QuerySingle{
+			Data: ConvertQuery(appl, ctx.Request, q),
+		}
+		ctx.ResponseData.Header().Set("Location", rest.AbsoluteURL(ctx.Request, app.LabelHref(ctx.SpaceID, res.Data.ID)))
```
The `LabelHref` should be replaced with `QueryHref`

pranavgore09 replied:
corrected in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/2c7f2d8b82dfd68ccfa159a7e6373bfe991266c4

baijum commented on controller/query.go:
```diff
@@ -0,0 +1,147 @@
+func ConvertQuery(appl application.Application, request *http.Request, q query.Query) *app.Query {
+	queryType := query.APIStringTypeQuery
+	spaceID := q.SpaceID.String()
+	relatedURL := rest.AbsoluteURL(request, app.LabelHref(spaceID, q.ID))
```
The `LabelHref` should be replaced with `QueryHref`

pranavgore09 replied:
corrected in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/2c7f2d8b82dfd68ccfa159a7e6373bfe991266c4

baijum commented on controller/query.go:
```diff
@@ -0,0 +1,147 @@
+func (c *QueryController) Show(ctx *app.ShowQueryContext) error {
+	// QueryController_Show: start_implement
+
+	// Put your logic here
```
This is required for `QueryHref` to work.

pranavgore09 replied:
yes that was todo part, added in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/2c7f2d8b82dfd68ccfa159a7e6373bfe991266c4

baijum commented on migration/sql-files/081-queries.sql:
```diff
@@ -0,0 +1,13 @@
+    deleted_at timestamp with time zone,
+    id uuid primary key DEFAULT uuid_generate_v4() NOT NULL,
+    title text NOT NULL CHECK(title <> ''),
+    fields jsonb,
```
Should be `NOT NULL`?

pranavgore09 replied:
yes that is a good thing to have, added in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/2c7f2d8b82dfd68ccfa159a7e6373bfe991266c4

baijum commented on test/testfixture/make_functions.go:
```diff
@@ -515,3 +516,35 @@ func makeTrackers(fxt *TestFixture) error {
+
+	for i := range fxt.Queries {
+		fxt.Queries[i] = &query.Query{
+			Title:   testsupport.CreateRandomValidTestName("label "),
```
Can we use a better word that `label ` here?

pranavgore09 replied:
haha `copy-paste` did that, not me :wink: 
Fixed in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/2c7f2d8b82dfd68ccfa159a7e6373bfe991266c4
thanks!

baijum commented on design/queries.go:
```diff
@@ -0,0 +1,104 @@
+	a.Attribute("created-at", d.DateTime, "When the query was created", func() {
+		a.Example("2016-11-29T23:18:14Z")
+	})
+	a.Attribute("updated-at", d.DateTime, "When the query was updated", func() {
```
We don't plan to support query update, right? If so, why do we need this `updated-at` attribute?

pranavgore09 replied:
yes right, removed in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/fc533a888888f670fba44bd208905e629ba34cda

baijum commented on design/queries.go:
```diff
@@ -0,0 +1,104 @@
+		a.Params(func() {
+			a.Param("queryID", d.UUID, "ID of the query")
+		})
+		// a.UseTrait("conditional")
```
You are going to enable this, right?

pranavgore09 replied:
Yeah I want to, but facing some error when I use traits.. I will check once agian

pranavgore09 replied:
facing following errror
```
/home/pranav/work/src/github.com/fabric8-services/fabric8-wit/app/conditional_requests.go:534:61: expected type, found '.'
/home/pranav/work/src/github.com/fabric8-services/fabric8-wit/app/conditional_requests.go:577:49: expected identifier

```

baijum commented on design/queries.go:
```diff
@@ -0,0 +1,104 @@
+			a.GET(""),
+		)
+		a.Description("List queries.")
+		// a.UseTrait("conditional")
```
You are going to enable this, right?

pranavgore09 replied:
same as https://github.com/fabric8-services/fabric8-wit/pull/1824#discussion_r158894294

pranavgore09 commented:
[test]

fabric8cd commented:
@pranavgore09 snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1824-9`

fabric8cd commented:
@pranavgore09 snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1824-12`

DhritiShikhar commented on design/queries.go:
```diff
@@ -0,0 +1,120 @@
+		a.Example("2016-11-29T23:18:14Z")
+	})
+	a.Attribute("fields", d.String, "Actual query string created by user", func() {
+		a.Example("...")
```
It would be good to have a valid query string here in the example.

DhritiShikhar commented on controller/query.go:
```diff
@@ -0,0 +1,190 @@
+	})
+}
+
+// Show runs the show action.
```
Please update this comment for Delete action

kwk replied:
I think it is fixed meanwhile, right? @DhritiShikhar

DhritiShikhar commented on design/queries.go:
```diff
@@ -0,0 +1,120 @@
+	a.Required("type", "attributes")
+})
+
+var queryRelationships = a.Type("QueryRelations", func() {
```
query relationships should contain space (e.g. https://github.com/fabric8-services/fabric8-wit/blob/master/design/labels.go#L46)

kwk commented on controller/query_blackbox_test.go:
```diff
@@ -0,0 +1,320 @@
+			// when
+			test.ListQueryUnauthorized(t, svc.Context, svc, ctrl, fxt.Spaces[0].ID, nil, nil)
+		})
+		t.Run("random sapce ID", func(t *testing.T) {
```
typo: `sapce` -> `space`. And Please rename to `unknown space ID`

kwk commented on controller/query_blackbox_test.go:
```diff
@@ -0,0 +1,320 @@
+	})
+}
+
+func (rest *TestQueryREST) TestShowQueryREST() {
```
Please rename to `TestShow`

kwk commented on controller/query_blackbox_test.go:
```diff
@@ -0,0 +1,320 @@
+			compareWithGoldenUUIDAgnostic(t, filepath.Join(rest.testDir, "show", "ok_show.res.query.golden.json"), queryObj)
+			compareWithGoldenUUIDAgnostic(t, filepath.Join(rest.testDir, "show", "ok_show.headers.golden.json"), resp.Header())
+		})
+		t.Run("ok without identity", func(t *testing.T) {
```
I'm not sure we should allow this use case.

kwk commented on controller/query_blackbox_test.go:
```diff
@@ -0,0 +1,320 @@
+			// when
+			test.ShowQueryNotFound(t, svc.Context, svc, ctrl, fxt2.Spaces[0].ID, fxt.Queries[0].ID, nil, nil)
+		})
+		t.Run("random space ID", func(t *testing.T) {
```
Please rename to `unknown space ID`

kwk commented on controller/query_blackbox_test.go:
```diff
@@ -0,0 +1,320 @@
+	})
+}
+
+func (rest *TestQueryREST) TestDeleteQueryREST() {
```
Please rename to `TestDelete`

kwk commented on design/queries.go:
```diff
@@ -0,0 +1,120 @@
+
+var queryAttributes = a.Type("QueryAttributes", func() {
+	a.Description(`JSONAPI store for all the "attributes" of a Query. See also http://jsonapi.org/format/#document-resource-object-attributes`)
+	a.Attribute("title", d.String, "The title of saved query", nameValidationFunction)
```
Please use `mandatoryOnCreate` here.

kwk commented on design/queries.go:
```diff
@@ -0,0 +1,120 @@
+	a.Attribute("created-at", d.DateTime, "When the query was created", func() {
+		a.Example("2016-11-29T23:18:14Z")
+	})
+	a.Attribute("fields", d.String, "Actual query string created by user", func() {
```
Please use `mandatoryOnCreate`

kwk commented on design/queries.go:
```diff
@@ -0,0 +1,120 @@
+
+var _ = a.Resource("query", func() {
+	a.Parent("space")
+	a.BasePath("/queries")
```
When you place this resource under `space` I suggest to also add a relationship link in the space response to find this endpoint naturally. I will look for it further down.

kwk commented on design/queries.go:
```diff
@@ -0,0 +1,120 @@
+		})
+		a.UseTrait("conditional")
+		a.Response(d.OK, querySingle)
+		a.Response(d.NotModified)
```
You don't have tests for the `not modified` response, do you?

kwk commented on migration/sql-test-files/081-query-empty-title.sql:
```diff
@@ -0,0 +1,9 @@
+-- empty query name
+-- this should fail
+DELETE FROM queries;
+INSERT INTO queries(fields, space_id, creator) 
```
You're not testing that the name is empty unless you pass it as `''`. In this test you check that is cannot be `NULL`.

kwk commented on query/query.go:
```diff
@@ -0,0 +1,169 @@
+	Fields  string
+}
+
+// GetETagData returns the field values to use to generate the ETag
```
This can be removed, right?

kwk commented on query/query.go:
```diff
@@ -0,0 +1,169 @@
+
+// GetETagData returns the field values to use to generate the ETag
+func (q Query) GetETagData() []interface{} {
+	return []interface{}{q.ID, strconv.FormatInt(q.UpdatedAt.Unix(), 10)}
```
Is there any particular reason to break with existing common practice of having a `Version` field and using that i for calculating the etag?

pranavgore09 replied:
we are not supporting `Update` on queries yet, so I have not introduced `Version` field on the table.
Hence I used updated_at time.
I see similar implementation for few other packages one of them is `Identity`

kwk replied:
Okay.

kwk commented on query/query.go:
```diff
@@ -0,0 +1,169 @@
+func (r *GormQueryRepository) Create(ctx context.Context, q *Query) error {
+	defer goa.MeasureSince([]string{"goa", "db", "Query", "create"}, time.Now())
+	q.ID = uuid.NewV4()
+	if strings.TrimSpace(q.Title) == "" {
```
We do enforce this on the DB layer as well. Isn't one check enough?

kwk commented on query/query.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+// List all labels in a space
+func (r *GormQueryRepository) List(ctx context.Context, spaceID uuid.UUID) ([]Query, error) {
```
Copy paste error: "labels" should be "queries". Also, what's the use case for listing all queries in a space? Isn't it much more useful to only list queries for a particular user?

kwk commented on query/query.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+// Repository describes interactions with Labels
+type Repository interface {
```
I think you should implement the `Exister` interface here.

kwk commented on query/query.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+// ListByCreator all labels in a space by a creator
+func (r *GormQueryRepository) ListByCreator(ctx context.Context, spaceID uuid.UUID, creatorID uuid.UUID) ([]Query, error) {
```
Copy past type: "labels" should be "queries". And yes, this is a valid requirement to me.

kwk commented on query/query.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+// Load Query in a space
+func (r *GormQueryRepository) Load(ctx context.Context, ID uuid.UUID, spaceID uuid.UUID) (*Query, error) {
```
Why do you need the space ID in order to load a query? Is the query's ID not "unique enough"? ;)

pranavgore09 replied:
hmm yes, I tried to align with other APIs hence made it space aware. it is really not required.
Will you suggest to replace spaceID by creatorID and add authentication to this API? user must be logged in and only he can load this particular object?

kwk replied:
This is the repo level. There should be no authentication. But the controller should double check that once loaded, the  query's author matches the current authenticated user.

kwk commented on query/query.go:
```diff
@@ -0,0 +1,169 @@
+
+// ListByCreator all labels in a space by a creator
+func (r *GormQueryRepository) ListByCreator(ctx context.Context, spaceID uuid.UUID, creatorID uuid.UUID) ([]Query, error) {
+	defer goa.MeasureSince([]string{"goa", "db", "Query", "list"}, time.Now())
```
This is not the "list" but the "listbycreator" function.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+	repo := query.NewQueryRepository(s.DB)
+	t.Run("success", func(t *testing.T) {
+		title := "My WI for sprin #101"
+		qs := `{"hello": "worold"}`
```
Not that it matters but "worold" should be "world" :)

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+		})
+		t.Run("invalid query json", func(t *testing.T) {
+			title := "My WI for sprin #101"
+			qs := "non-json query"
```
I couldn't not find a place in the repo's creation function that checks for valid query JSON. Can you point me to the correct location, please?

pranavgore09 replied:
`fields` column is **jsonb** and hence it accepts only valid json. nothing to be done more for that IMO hence no code in the repo for that.

kwk replied:
I do think there is a need to check for valid query syntax. See my answer to this https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/8936aa1ff1018e648fff1a0e66be3355a8eef093#r159368461

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+			for _, q := range qList {
+				delete(mustHave, q.Title)
+			}
+			assert.Empty(s.T(), mustHave)
```
You must not use `s.T()` here but `t`.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+		t.Run("by spaceID", func(t *testing.T) {
+
+			// given
+			fxt := tf.NewTestFixture(s.T(), s.DB,
```
You must not use `s.T()` here but `t`.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+		title := "My WI for sprin #101"
+		qs := `{"hello": "worold"}`
+		// given
+		fxt := tf.NewTestFixture(s.T(), s.DB, tf.Spaces(1))
```
You must not use `s.T()` here but `t`.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+			title := ""
+			qs := `{"hello": "worold"}`
+			// given
+			fxt := tf.NewTestFixture(s.T(), s.DB, tf.Spaces(1))
```
You must not use `s.T()` here but `t`.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+			title := "My WI for sprin #101"
+			qs := "non-json query"
+			// given
+			fxt := tf.NewTestFixture(s.T(), s.DB, tf.Spaces(1))
```
You must not use `s.T()` here but `t`.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+		})
+		t.Run("by spaceID and creator", func(t *testing.T) {
+			// given
+			fxt := tf.NewTestFixture(s.T(), s.DB,
```
You must not use `s.T()` here but `t`.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+			for _, q := range qList {
+				delete(mustHave, q.Title)
+			}
+			assert.Empty(s.T(), mustHave)
```
You must not use `s.T()` here but `t`.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+	repo := query.NewQueryRepository(s.DB)
+	t.Run("success", func(t *testing.T) {
+		// given
+		fxt := tf.NewTestFixture(s.T(), s.DB,
```
You must not use `s.T()` here but `t`.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+		q, err := repo.Load(context.Background(), fxt.QueryByTitle("q1").ID, fxt.Spaces[0].ID)
+		// then
+		require.Nil(t, err)
+		assert.Equal(s.T(), "q1", q.Title)
```
You must not use `s.T()` here but `t`.

kwk replied:
This is not fixed yet.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+	})
+}
+
+func (s *TestQueryRepository) TestShowQuery() {
```
Please rename to `TestShow`.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+
+}
+
+func (s *TestQueryRepository) TestListQuery() {
```
Please rename to `TestList`

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+	suite.Run(t, &TestQueryRepository{DBTestSuite: gormtestsupport.NewDBTestSuite("../config.yaml")})
+}
+
+func (s *TestQueryRepository) TestCreateQuery() {
```
Please rename to `TestCreate`.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+	repo := query.NewQueryRepository(s.DB)
+	t.Run("success", func(t *testing.T) {
+		// given
+		fxt := tf.NewTestFixture(s.T(), s.DB,
```
You must not use `s.T()` here but `t`.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+	})
+}
+func (s *TestQueryRepository) TestDeleteQuery() {
+	t := s.T()
```
Please remove this.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+}
+
+func (s *TestQueryRepository) TestShowQuery() {
+	t := s.T()
```
Please remove this.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+}
+
+func (s *TestQueryRepository) TestListQuery() {
+	t := s.T()
```
Please remove this.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+		// when
+		err := repo.Delete(context.Background(), fxt.QueryByTitle("q1").ID)
+		// then
+		require.Nil(t, err)
```
Please use `require.NoError(t, err)` instead.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+	})
+	t.Run("fail", func(t *testing.T) {
+		_, err := repo.Load(context.Background(), uuid.NewV4(), uuid.NewV4())
+		require.NotNil(t, err)
```
Please use `require.Error(t, err)` instead.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+		// when
+		q, err := repo.Load(context.Background(), fxt.QueryByTitle("q1").ID, fxt.Spaces[0].ID)
+		// then
+		require.Nil(t, err)
```
Please use `require.NoError(t, err)` instead.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+			// when
+			qList, err := repo.ListByCreator(context.Background(), fxt.Spaces[0].ID, fxt.Identities[0].ID)
+			// then
+			require.Nil(t, err)
```
Please use `require.NoError(t, err)` instead.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+			// when
+			qList, err := repo.List(context.Background(), fxt.Spaces[0].ID)
+			// then
+			require.Nil(t, err)
```
Please use `require.NoError(t, err)` instead.

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,168 @@
+			// when
+			err := repo.Create(context.Background(), &q)
+			// then
+			require.NotNil(t, err)
```
Please use `require.Error(t, err)` instead.

DhritiShikhar commented on controller/query.go:
```diff
@@ -0,0 +1,190 @@
+		return jsonapi.JSONErrorResponse(ctx, goa.ErrUnauthorized(err.Error()))
+	}
+	return application.Transactional(c.db, func(appl application.Application) error {
+		_, err = appl.Spaces().Load(ctx, ctx.SpaceID)
```
Shouldn't you check for conditions like 'if ctx.SpaceID == nil' before using it?

kwk replied:
I think that `ctx.SpaceID` is not a pointer but a normal member.

kwk commented on test/testfixture/make_functions.go:
```diff
@@ -515,3 +516,35 @@ func makeTrackers(fxt *TestFixture) error {
+	for i := range fxt.Queries {
+		fxt.Queries[i] = &query.Query{
+			Title:   testsupport.CreateRandomValidTestName("query "),
+			Fields:  `{"key": "value"}`,
```
I highly doubt that this is a valid query expression and it probably only works because there's no check in the query repo's creation function.

kwk commented on query/query.go:
```diff
@@ -0,0 +1,169 @@
+const QueryTableName = "queries"
+
+// Create a new label
+func (r *GormQueryRepository) Create(ctx context.Context, q *Query) error {
```
Typo: "label" should be "query". Also, I'm missing a call to `ParseFilterString` in this function. It blindly accepts anything that is valid JSON it seems.

pranavgore09 replied:
You are right but I wanted repository unaware of application logic, so I am using  `ParseFilterString` in the controller before save to validate syntax & keys.

kwk replied:
The query itself is not handled in application logic but also on repository level. Therefore your intention does not hold true. I think we should do data validation on repo level.

DhritiShikhar commented on controller/query.go:
```diff
@@ -0,0 +1,190 @@
+		return jsonapi.JSONErrorResponse(ctx, goa.ErrUnauthorized(err.Error()))
+	}
+	err = application.Transactional(c.db, func(appl application.Application) error {
+		q, err := appl.Queries().Load(ctx.Context, ctx.QueryID, ctx.SpaceID)
```
please check if space exists

DhritiShikhar commented on controller/query_blackbox_test.go:
```diff
@@ -0,0 +1,320 @@
+			svc, ctrl = rest.SecuredControllerWithIdentity(fxt2.Identities[0])
+			_, qList = test.ListQueryOK(t, svc.Context, svc, ctrl, fxt2.Spaces[0].ID, nil, nil)
+			// then
+			require.NotNil(t, qList)
```
@pranavgore09 you said user should be able to see his own saved queries only. Then shouldn't your test case -> list by different user return a `nil` `qList`?

pranavgore09 commented:
@DhritiShikhar @kwk I have addressed your comments. Thanks for such thorough review!
Question:
1. Should we keep SHOW endpoint for all users? (without login)
2. Should we allow the user to create an exact same query with the same title at the same time?

fabric8cd commented:
@pranavgore09 snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1824-14`

kwk commented:
@pranavgore09 

> Should we keep SHOW endpoint for all users? (without login)

No, only the user who created the query should be able to view it IMHO.

> Should we allow the user to create an exact same query with the same title at the same time?

No, we should limit it with the `deleted_at` exception.

DhritiShikhar commented:
@pranavgore09 thank you so much for raising those two questions.

1. I think we should keep the `show`endpoint with login.
The `show` endpoint for custom query is required because there could be some case when we just want a single custom query. 

But I feel, it should be visible to the creator only because as of now, custom queries are only for users, right? So why have separate access control for list and show? 

2. This is not possible. Is it? There would be some time time difference, nanoseconds or something. Anyway, we should avoid this scenario and add checks so that same user cannot create an exact same query with the same title at the same time

kwk commented on migration/sql-test-files/081-query-null-title.sql:
```diff
@@ -0,0 +1,9 @@
+DELETE FROM spaces where id='171d52ff-fa00-46d7-ac24-94269908ad7a';
+INSERT INTO spaces (id, name) VALUES ('171d52ff-fa00-46d7-ac24-94269908ad7a', 'test space');
+
+-- empty query name
```
In this test, the name is not empty as in `''` but it is `NULL`. Therefore you're testing that Postgres does it job when checking for `NOT NULL` values. I think we don't need to go this deep.

pranavgore09 replied:
yes, hence the filename. and added https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/8936aa1ff1018e648fff1a0e66be3355a8eef093#diff-06f8f427154426f3c99f95b5424025c7 to see the empty title
do you suggest I should not have this test NULL title?

kwk replied:
yes

kwk replied:
the empty title is okay but we should rely on postgres to check that the title cannot be null. Also, if there's a postgres check on the title that it cannot bet `''`, then I think you can get rid of that test for the empty title alltogether. WDYT?

pranavgore09 replied:
sure, I will get rid off null_title test thanks!

kwk commented on query/query_test.go:
```diff
@@ -0,0 +1,163 @@
+		q, err := repo.Load(context.Background(), fxt.QueryByTitle("q1").ID, fxt.Spaces[0].ID)
+		// then
+		require.NoError(t, err)
+		assert.Equal(s.T(), "q1", q.Title)
```
Please use `t` here.

fabric8cd commented:
@pranavgore09 snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1824-15`

pranavgore09 commented:
@kwk aah! I tried checking query at repository level and facing cyclic import
```
# github.com/fabric8-services/fabric8-wit/search
import cycle not allowed in test
package github.com/fabric8-services/fabric8-wit/search (test)
	imports github.com/fabric8-services/fabric8-wit/test/testfixture
	imports github.com/fabric8-services/fabric8-wit/query
	imports github.com/fabric8-services/fabric8-wit/search

FAIL	github.com/fabric8-services/fabric8-wit/search [setup failed]
```

fabric8cd commented:
@pranavgore09 snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1824-18`

fabric8cd commented:
@pranavgore09 snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1824-19`

fabric8cd commented:
@pranavgore09 snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1824-20`

fabric8cd commented:
@pranavgore09 snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1824-21`

baijum commented on controller/query.go:
```diff
@@ -0,0 +1,210 @@
+	"github.com/fabric8-services/fabric8-wit/rest"
+	"github.com/fabric8-services/fabric8-wit/space"
+	"github.com/goadesign/goa"
+	"github.com/prometheus/common/log"
```
Wrong import path.

pranavgore09 replied:
thanks @baijum   fixed in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/50108ded08fe6e3b3a39474e319cf43c6c30f98b

fabric8cd commented:
@pranavgore09 snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1824-22`

fabric8cd commented:
@pranavgore09 snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1824-23`

fabric8cd commented:
@pranavgore09 snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1824-24`

kwk commented on controller/query.go:
```diff
@@ -0,0 +1,210 @@
+			CreatedAt: &q.CreatedAt,
+		},
+		Links: &app.GenericLinks{
+			Self:    &relatedURL,
```
You might as well use the `ptr.String` function to get pointer values on-the-fly.

pranavgore09 replied:
I am using ptr package here https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/c234bcd1f0b1fe7131b360c25bd9707386f3ede5
but is it okay to use a variable if I am using it more than once?

kwk replied:
Sure, then I would use a variable too. +1

kwk commented on controller/query.go:
```diff
@@ -0,0 +1,190 @@
+		return jsonapi.JSONErrorResponse(ctx, goa.ErrUnauthorized(err.Error()))
+	}
+	return application.Transactional(c.db, func(appl application.Application) error {
+		_, err = appl.Spaces().Load(ctx, ctx.SpaceID)
```
@pranavgore09 do you have an answer to that comment meanwhile?

pranavgore09 replied:
I am already using `Exister` for create, and now doing the same in `list` as well https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/b7ae234e2887a5facf8b50ea655aa7dfa39996b2

kwk commented on controller/query_blackbox_test.go:
```diff
@@ -0,0 +1,347 @@
+				"q6": {},
+			}
+			for _, q := range qList.Data {
+				delete(mustHave, q.Attributes.Title)
```
Hey, I just [realized](https://golang.org/pkg/builtin/#delete) that you can call `delete` before checking if the key exists. Great.

kwk commented on controller/query_blackbox_test.go:
```diff
@@ -0,0 +1,347 @@
+		t.Run("unknown space ID", func(t *testing.T) {
+			// given
+			fxt := tf.NewTestFixture(t, rest.DB, tf.CreateWorkItemEnvironment())
+			svc, ctrl := rest.SecuredControllerWithIdentity(fxt.Identities[0])
```
Since you're only using the identity I suggest you only create an Identity in the `fxt` setup above.

pranavgore09 replied:
sure! done in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/b7ae234e2887a5facf8b50ea655aa7dfa39996b2

kwk commented on query/query.go:
```diff
@@ -0,0 +1,202 @@
+	return &q, nil
+}
+
+// IsDuplicate checks if there is already a query with same content by same user
```
@pranavgore09 why not simply insert and check for a duplicate with [IsUniqueViolation`](https://github.com/fabric8-services/fabric8-wit/blob/master/gormsupport/postgres.go#L38) ? If you want to have on-key-press-validation in the UI, you will need this function of course.

pranavgore09 replied:
Yes, @kwk I tried that. Our unique condition includes `deleted_at`.
Let's say that there is a record already present with title1, space1, and crreator1
If I am trying to create the same record, DB accepts it and creates a row and we get 201.
But I am testing data conflict error here https://github.com/fabric8-services/fabric8-wit/pull/1824/files#diff-5167477ec144512fe6c115235dae2ca8R162 and without calling `IsDuplicate` manually I do not get an error from DB.

I was expecting the behavior of code same as you were. So there is something wrong in design https://github.com/fabric8-services/fabric8-wit/pull/1824/files#diff-69e4ee5c520dbf4a8b435472706ab68dR10 or something wrong with my understanding.

kwk replied:
Hm, if the unique condition includes the `deleted_at` column you should be able to create query "foo", delete it and recreate it again without a name clash. Only if there exists a query foo that is not deleted, recreation should be prohibited. I'm checking the query.

pranavgore09 replied:
fixed `index` in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/bbbd0b4dd1d3d1fd9266c87bda23c26d5a7d67a7
removed `IsDuplicate` in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/ca299ba46d036f827341c44276bc0282ff75759a
thanks @kwk

kwk commented on query/query.go:
```diff
@@ -0,0 +1,202 @@
+// Delete deletes the query with the given id, returns NotFoundError or InternalError
+func (r *GormQueryRepository) Delete(ctx context.Context, ID uuid.UUID) error {
+	defer goa.MeasureSince([]string{"goa", "db", "query", "delete"}, time.Now())
+	if ID == uuid.Nil {
```
Why treat `uuid.Nil` as a different value instead of just passing it on?

pranavgore09 replied:
true that, fixed and added a test in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/a79d9d40df2ccb905dfd34647bf91b4d0d76482a

pranavgore09 commented:
[test]

fabric8cd commented:
@pranavgore09 snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1824-25`

fabric8cd commented:
@pranavgore09 snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1824-26`

fabric8cd commented:
@pranavgore09 snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1824-27`

kwk commented on query/query.go:
```diff
@@ -0,0 +1,172 @@
+	if tx.RowsAffected == 0 {
+		log.Error(ctx, map[string]interface{}{
+			"query_id": ID.String(),
+		}, "none row was affected by the delete operation")
```
typo: "none" should be "no"

pranavgore09 replied:
thanks fixed in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/354050c56766743806b6eaeb78751e72c7546d95

kwk commented on test/testfixture/make_functions.go:
```diff
@@ -515,3 +516,35 @@ func makeTrackers(fxt *TestFixture) error {
+	for i := range fxt.Queries {
+		fxt.Queries[i] = &query.Query{
+			Title:   testsupport.CreateRandomValidTestName("query "),
+			Fields:  `{"space": "a2bda47f-4c4a-400c-9e30-62b11eaf8900"}`,
```
This needs to go away I'd say. Can you initialize with no value for `Fields` and do something like this instead?

```go
if !fxt.IsolatedCreation {
    fxt.Queries[i].Fields= fmt.Sprintf(`{"space": "%s"}`, fxt.Spaces[0].ID)
    fxt.Queries[i].SpaceID = fxt.Spaces[0].ID
}
```

pranavgore09 replied:
yes done in https://github.com/fabric8-services/fabric8-wit/pull/1824/commits/354050c56766743806b6eaeb78751e72c7546d95

fabric8cd commented:
@pranavgore09 snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1824-28`

pranavgore09 commented on controller/query.go:
```diff
@@ -0,0 +1,206 @@
+		return jsonapi.JSONErrorResponse(ctx, goa.ErrUnauthorized(err.Error()))
+	}
+	return application.Transactional(c.db, func(appl application.Application) error {
+		err = appl.Spaces().CheckExists(ctx, ctx.SpaceID)
```
it helps in returning proper `NOT FOUND` error instead of `INTERNAL SERVER ERROR`.
Do you suggest to return `500` for unknown space IDs ?

fabric8cd commented:
@pranavgore09 snapshot fabric8-wit image is available for testing.  `docker pull docker.io/fabric8/fabric8-wit:SNAPSHOT-PR-1824-29`
